### DISCUSSION TITLE
Faraday 2.0 Alpha

### DISCUSSION THREAD

**Discussion by iMacTia** [lostisland/faraday#1337] [General]
## Faraday 2.0.0.alpha-1

This is the first alpha release for the upcoming Faraday 2.0 major release 🎉 .
Please note, this version is not production ready, but we'd appreciate everyone feedback if you have the chance to try it in your personal projects 🙌.

To learn more about how to safely upgrade to Faraday 2.0 and update your code, please checkout our [UPGRADING](https://github.com/lostisland/faraday/blob/main/UPGRADING.md) guide!

And you have any feedback, please feel free to join us in Discussions 💬 !

<hr /><em>This discussion was created from the release <a href='https://github.com/lostisland/faraday/releases/tag/v2.0.0.alpha-1'>v2.0.0.alpha-1</a>.</em>

**Comment by mattbrictson** [comment]
Hi! Thanks for all your work on getting Faraday 2.0 ready.

I'm wanting to try 2.0.0.alpha.pre.1 on a personal project, as suggested in the release notes. But my project also requires `faraday_middleware`. I don't see any way to use faraday_middleware in conjunction with faraday >= 2.0.0.alpha.pre.1 because the gemspec is incompatible: it is pinned to ~> 1.0.

https://github.com/lostisland/faraday_middleware/blob/61b1422c831434caa1d84ab35e2e77f1ada39679/faraday_middleware.gemspec#L19

Any suggestions?

**Reply by mattbrictson** [reply]
Oh! I just found that the `json` middleware has been moved into the core `faraday` gem: https://github.com/lostisland/faraday/pull/1300. That is big news because it is the whole reason I have the `faraday_middleware` dependency. Now that it is in core, I don't need the extra gem.

Maybe this should be mentioned in the UPGRADING guide? I suspect many developers will be in the same boat.

**Reply by iMacTia** [reply]
That's a really great point, thank you for the feedback @mattbrictson 🙌 !
It's not a strict compatibility requirement, that's why I didn't included it in the UPGRADING file and was instead leaving that for the release notes, but you got a point on the interoperability with `faraday_middleware`, so it should definitely be mentioned 👍 !

Thank you so much for trying the 2.0 alpha and providing feedback, this is invaluable for us 🙇 !

**Comment by mattbrictson** [comment]
I am still a bit stuck on how to use the 2.0 alpha.

It looks like the `adapter` setting is being ignored. Here is a simple script:

```ruby
require "bundler/inline"

gemfile do
  source "https://rubygems.org"
  gem "faraday", "2.0.0.alpha.pre.1"
  gem "faraday-net_http"
end

conn = Faraday.new do |builder|
  builder.adapter :net_http
end

conn.get("https://github.com/")
```

This blows up:

```
gems/faraday-2.0.0.alpha.pre.1/lib/faraday/middleware_registry.rb:57:in `lookup_middleware': :test is not registered on Faraday::Adapter (Faraday::Error)
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday/rack_builder.rb:114:in `adapter'
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday/rack_builder.rb:77:in `build'
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday/rack_builder.rb:65:in `initialize'
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday/options/connection_options.rb:19:in `new'
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday/options/connection_options.rb:19:in `new_builder'
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday/connection.rb:81:in `initialize'
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday.rb:96:in `new'
  from gems/faraday-2.0.0.alpha.pre.1/lib/faraday.rb:96:in `new'
  from faraday.rb:9:in `<main>'
```

Digging through the faraday source, I see that it falls back to an adapter named `:test` if no adapter is specified. But then `:test` is apparently an invalid adapter name, hence the exception.

A couple points of confusion here:

1. Why is faraday ignoring `builder.adapter :net_http`? Do I have that written incorrectly?
2. If I am using the builder API wrong, why am I not getting an `ArgumentError`?
3. `:test is not registered on Faraday::Adapter` is a very confusing error for someone not intimately familiar with the faraday source code. Could this be improved for users who forget to specify an adapter?

**Reply by mattbrictson** [reply]
It looks like the block passed to `Faraday.new` is not being evaluated at all. For example, the `raise` in this code is never reached:

```ruby
conn = Faraday.new do |builder|
  raise "wat"
  builder.adapter :net_http
end
```

**Reply by iMacTia** [reply]
Hi @mattbrictson and thanks once more for testing the alpha!
I was working on this last weekend and I noticed that we have a somewhat complicated setup in the rack builder where a stack is first created without any config (hence defaulting to the default adapter) and only AFTER this, the actual block you provide is called.
This normally works because the custom block will simply override the default configuration, but I agree it should be refactored at some point. (if interested, [here](https://github.com/lostisland/faraday/blob/main/lib/faraday/connection.rb#L81) is where we do that)

Since you get an exception during the first step, then the code won't reach your block, explaining why you don't see the exception.
Long term solution would be to improve the rack builder, but since I'm aiming for a major rewrite after 2.0, I'd just fix the issue with the test adapter for now and that should solve the problem.

I believe the actual issue we're looking at has to do with loading, I'll take a look immediately

**Reply by iMacTia** [reply]
As I thought, the test adapter is not required anywhere except during `rspec`. I just pushed a [commit](https://github.com/lostisland/faraday/commit/498f6027a18df15561723da8d6cb10288a24bf64) to fix that, and released `alpha-2`. Your code works for me replacing the faraday version with `2.0.0.alpha.pre.2` 🙌

**Reply by mattbrictson** [reply]
Great! Confirmed it is now working with `2.0.0.alpha.pre.2`. Thanks!

**Comment by mattbrictson** [comment]
One more question: what is the status of `faraday-net_http_persistent`?

If I try this code:

```ruby
require "bundler/inline"

gemfile do
  source "https://rubygems.org"
  gem "faraday", "2.0.0.alpha.pre.2"
  gem "faraday-net_http_persistent"
end

conn = Faraday.new do |builder|
  builder.adapter :net_http_persistent
end

conn.get("https://github.com/")
```

Then I get this exception:

```
gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/adapter/net_http_persistent.rb:6:in `<class:Adapter>': uninitialized constant Faraday::Adapter::NetHttp (NameError)
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/adapter/net_http_persistent.rb:4:in `<module:Faraday>'
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/adapter/net_http_persistent.rb:3:in `<top (required)>'
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/net_http_persistent.rb:3:in `require_relative'
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/net_http_persistent.rb:3:in `<top (required)>'
  from site_ruby/3.0.0/bundler/runtime.rb:73:in `require'
  from site_ruby/3.0.0/bundler/runtime.rb:73:in `rescue in block in require'
  from site_ruby/3.0.0/bundler/runtime.rb:51:in `block in require'
  from site_ruby/3.0.0/bundler/runtime.rb:44:in `each'
  from site_ruby/3.0.0/bundler/runtime.rb:44:in `require'
  from site_ruby/3.0.0/bundler/inline.rb:71:in `block in gemfile'
  from site_ruby/3.0.0/bundler/settings.rb:131:in `temporary'
  from site_ruby/3.0.0/bundler/inline.rb:55:in `gemfile'
  from faraday.rb:3:in `<main>'
```

So it looks like there is an implicit, undeclared dependency on the `faraday-net_http` gem?

If I then explicitly add `gem "faraday-net_http"` to my gemfile, then I get a different error:

```
gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/adapter/net_http_persistent.rb:7:in `<class:NetHttpPersistent>': undefined method `dependency' for Faraday::Adapter::NetHttpPersistent:Class (NoMethodError)
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/adapter/net_http_persistent.rb:6:in `<class:Adapter>'
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/adapter/net_http_persistent.rb:4:in `<module:Faraday>'
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/adapter/net_http_persistent.rb:3:in `<top (required)>'
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/net_http_persistent.rb:3:in `require_relative'
  from gems/3.0.0/gems/faraday-net_http_persistent-1.2.0/lib/faraday/net_http_persistent.rb:3:in `<top (required)>'
  from site_ruby/3.0.0/bundler/runtime.rb:73:in `require'
  from site_ruby/3.0.0/bundler/runtime.rb:73:in `rescue in block in require'
  from site_ruby/3.0.0/bundler/runtime.rb:51:in `block in require'
  from site_ruby/3.0.0/bundler/runtime.rb:44:in `each'
  from site_ruby/3.0.0/bundler/runtime.rb:44:in `require'
  from site_ruby/3.0.0/bundler/inline.rb:71:in `block in gemfile'
  from site_ruby/3.0.0/bundler/settings.rb:131:in `temporary'
  from site_ruby/3.0.0/bundler/inline.rb:55:in `gemfile'
  from faraday.rb:3:in `<main>'
```

It is not clear to me what that means. Any idea what is causing this? Perhaps `faraday-net_http_persistent` is not compatible with 2.0 yet?

**Reply by olleolleolle** [reply]
dependency is the  Faraday 1.x way of attempting to manage the loading of optional dependencies. 

Yeah, an adapter that needs 2.x support, I think.

**Reply by mattbrictson** [reply]
Is it fair to say that instead of using `dependency`, adapters should be changed to explicitly list their dependencies in their `gemspec`? In Faraday 1.x, I understand that `dependency` was needed because all the adapters existed in the core gem, so we needed a way to let people opt into certain adapters. However in 2.x, if I want to use a particular adapter, I have to add that adapter to my Gemfile. So it seems like any dependencies of that adapter should just automatically come with it.

That would mean existing adapters will need a breaking change in order to become compatible with 2.x.

Is that what you had in mind?

Let me know if I can help. I am happy to open a PR for `faraday-net_http_persistent` as long as we're on the same page in terms of how things should generally proceed.

**Reply by iMacTia** [reply]
YES, that's the whole point of separating adapters out, so that we don't have dependencies mess anymore 🎉
You're spot-on @mattbrictson 🙌

And yes some adapters (if not all), will likely need an update to fully support Faraday 2.0.

I'm sorry I didn't have time to follow up on that yet, but now that we have the 2.0 version out, I can move forward with those as well

**Reply by iMacTia** [reply]
I have a PR open that fixes this issue for `net_http` now: https://github.com/lostisland/faraday-net_http/pull/9
I'll go through the other adapters as well, but I don't think I'll update them all before the final v2.0 is released.
My approach would be to proceed with a community interest-based priority

**Reply by iMacTia** [reply]
@mattbrictson PR merged and `faraday-net_http` v2.0.0.alpha-1 have been released 🎉

**Reply by mattbrictson** [reply]
I also opened this PR to make `faraday-net_http_persistent` compatible with 2.0: https://github.com/lostisland/faraday-net_http_persistent/pull/7

**Reply by iMacTia** [reply]
Ahaha sorry @mattbrictson I totally missed that and opened one myself 🙈 https://github.com/lostisland/faraday-net_http_persistent/pull/8

I like the changes you made to the README as well, and it seems like you kept `net-http-persistent` v3 compatibility, which I didn't.
Feel free to cherry-pick any other change from my PR (which I'm going to close now). Any further commit you do to your branch will now be built as I've reactivated the CI workflow 👍

**Reply by olleolleolle** [reply]
@mattbrictson Thanks so much for making that PR, we are rolling!
